presto-common/src/main/java/com/facebook/presto/common/block/DictionaryBlock.java

```diff
@@ -264,14 +264,25 @@ public long getRegionLogicalSizeInBytes(int positionOffset, int length)
         }
 
         long sizeInBytes = 0;
-        long[] seenSizes = new long[dictionary.getPositionCount()];
-        Arrays.fill(seenSizes, -1L);
-        for (int i = positionOffset; i < positionOffset + length; i++) {
-            int position = getId(i);
-            if (seenSizes[position] < 0) {
-                seenSizes[position] = dictionary.getRegionLogicalSizeInBytes(position, 1);
+        // Dictionary Block may contain large number of keys and small region length may be requested.
+        // If the length is less than keys the cache is likely to be not used.
+        if (length > dictionary.getPositionCount()) {
+            // Cache code path.
+            long[] seenSizes = new long[dictionary.getPositionCount()];
+            Arrays.fill(seenSizes, -1L);
+            for (int i = positionOffset; i < positionOffset + length; i++) {
+                int position = getId(i);
+                if (seenSizes[position] < 0) {
+                    seenSizes[position] = dictionary.getRegionLogicalSizeInBytes(position, 1);
+                }
+                sizeInBytes += seenSizes[position];
```

Comment:
Q regarding the original implementation. This is a dictionary, do we really need to multiply the entry size by the number of times it appears?

If the answer is no, we can try to collect used dictionary positions into a IntOpenHashSet from fastutil. 

If yes, instead of a regular hash map you can use Int2IntOpenHashMap. It would really depend on the relation between the dict size and the length, and still likely be more expensive.

Reply:
The logical size, wants to get the underlying size of the data. That will involve multiplying the entry size, by number of times it appears. The logical size is used in few places, where the data is converted from dictionary encoding to direct encoding. One such place is abandoning dictionary encoding and switching to direct encoding.

I agree that fastutil's collections are better than java.utils in terms of CPU/memory. But my opinion, is the code is simple when the likelihood of hit is very low, we can have a non cached path and avoid the cache completely.